src/jltypes.c

```diff
@@ -2866,6 +2866,7 @@ JL_DLLEXPORT jl_svec_t *jl_compute_fieldtypes(jl_datatype_t *st JL_PROPAGATES_RO
     top.prev = (jl_typestack_t*)stack;
     st->types = inst_ftypes(wt->types, &env[n - 1], &top, cacheable);
     jl_gc_wb(st, st->types);
+    jl_compute_has_concrete_subtype(st);
```

Comment:
I preferred to call this here after computing field types, rather than moving it to somewhere here: https://github.com/JuliaLang/julia/blob/b9ac28a645fee63f08a607e8b66f7b75bb05159d/src/datatype.c#L349-L359 because `st->types` may be `NULL`.

I can investigate whether a simple `st->types != NULL` guard in `jl_struct_try_layout` would be enough, but I thought it might make more sense to do it here in `jl_compute_fieldtypes` so we are sure that it gets cached at the same time as field computations, as early as possible.